### PR TITLE
Relog configuration args on debug.log rotation

### PR DESCRIPTION
**Pull request by LarryRuane**
This PR was part of #16115 but reviewers requested that this be made into a separate PR. As explained in #16115, it's useful, maybe even crucial, for support, analysis, and debugging purposes for the configuration arguments to be present in the `debug.log` file, as that may be the only artifact provided to the developer or support team. That PR logs the configuration settings (command-line arguments and `bitcoin.conf` contents) when `bitcoind` starts. But sometime later, the debug log file may be "rotated" (a new `debug.log` file started so it doesn't grow unbounded in size) and this information would be lost unless the previous `debug.log` is preserved.

This PR re-logs the configuration information when the log rotates, ensuring that the current `debug.log` file contains this information. Reviewers had reservations about the previous approach (which you can see in #16115's discussion thread), so this PR implements this functionality in a simpler and (I believe) much better way.

### CODE REVIEW

File: test/functional/test_framework/test_node.py [hunk 1]
```diff
@@ -307,7 +307,7 @@ def wait_until_stopped(self, timeout=BITCOIND_PROC_WAIT_TIMEOUT):
         wait_until(self.is_node_stopped, timeout=timeout)
 
     @contextlib.contextmanager
-    def assert_debug_log(self, expected_msgs, timeout=2):
+    def assert_debug_log(self, expected_msgs, unexpected_msgs=[], timeout=2):
```
Comment (practicalswift) [inline test/functional/test_framework/test_node.py]: Do you really want `unexpected_msgs` to preserve parameter state between calls? :-)

```
>>> def foobar(foo, bar=[]):
...     bar.append(foo)
...     print(bar)
...
>>> foobar("hello")
['hello']
>>> foobar("hello")
['hello', 'hello']
>>> foobar("hello")
['hello', 'hello', 'hello']
```

FWIW I think we should use a linter to guard against this Python gotcha :-)

Comment (LarryRuane) [reply test/functional/test_framework/test_node.py]: Great catch, I had no idea! I fixed this in #16115 (where it was introduced) and force-pushed rebased this PR onto that branch.

Comment (LarryRuane) [reply test/functional/test_framework/test_node.py]: @practicalswift, I don't think this was an actual problem. I added a debug print and `unexpected_msgs` didn't include previous values. I think it's not broken because this function, `assert_debug_log()`, doesn't modify the variable (it's not mutable). Here's an extremely simplified version of this function and some sample calls to it:
```
def assert_debug_log(unexpected_msgs=[]):
    print(unexpected_msgs)
    for unexpected_msg in unexpected_msgs:
        print('  ', unexpected_msg)

assert_debug_log()
assert_debug_log(["foo"])
assert_debug_log(unexpected_msgs=["foo"])
assert_debug_log("bar") # not actually called this way but just curious
assert_debug_log()
```
The output:
```
[]
['foo']
   foo
['foo']
   foo
bar
   b
   a
   r
[]
```
What prompted me to investigate this further is that I wondered if any of our existing python code has this problem. I found just one instance, which is in `wallet_importmulti.py`:
```
def test_importmulti(self, req, success, error_code=None, error_message=None, warnings=[]):
```
But adding a debug print to that function showed everything was okay. I believe that's because `warnings` isn't modified in that function.

**Comment by LarryRuane** [comment]
Rebased onto https://github.com/bitcoin/bitcoin/pull/16115 (only the last commit here, 339a1e3, is unique to this PR; the previous commits are part of https://github.com/bitcoin/bitcoin/pull/16115)

File: src/logging.cpp [hunk 2]
```diff
@@ -282,9 +286,6 @@ void BCLog::Logger::LogPrintStr(const std::string& str)
         fwrite(str_prefixed.data(), 1, str_prefixed.size(), stdout);
         fflush(stdout);
     }
-    for (const auto& cb : m_print_callbacks) {
-        cb(str_prefixed);
-    }
     if (m_print_to_file) {
```
Comment (LarryRuane) [inline src/logging.cpp]: It may be better to have entries for `m_print_to_file` (and also `m_print_to_console` above) in the `m_print_callbacks` list, rather than having them here as special cases.  I may investigate that approach, and if it seems reasonable, make a separate PR (to merge ahead of the current PR).

File: src/logging.cpp [hunk 3]
```diff
@@ -296,10 +297,27 @@ void BCLog::Logger::LogPrintStr(const std::string& str)
...
+
+                // Re-log important messages to the new file.
+                bool save_started_new_line = m_started_new_line;
+                m_started_new_line = true;
+                for (const std::string& msg : m_msgs_relog) {
+                    std::string s = AddPrefix(msg) + '\n';
+                    FileWriteStr(s, m_fileout);
+                    for (const auto& cb : m_print_callbacks) {
+                        cb(s);
+                    }
```
Comment (LarryRuane) [inline src/logging.cpp]: It's bad that this loop is duplicated here, and weird too, because the various unknown future print callbacks probably don't care that the log was rotated. But this is necessary for the test (which uses the callback mechanism). It may be better for the test to simply open the real `debug.log` file, as it used to, even though that's not quite as nice and clean.

**Review by LarryRuane — COMMENTED** [review]
Force-pushed after rebasing onto https://github.com/bitcoin/bitcoin/pull/16115

**Comment by LarryRuane** [comment]
Rebased onto latest master

**Comment by LarryRuane** [comment]
Force-pushed merge conflict fixes. I believe all reviewer comments have been addressed; no more work planned, this PR can be merged if desired.

**Comment by jonatack** [comment]
Concept ACK. I'm a fan of the original feature that logs the config options to the debug log.

**Comment by LarryRuane** [comment]
Force-pushed required rebase. I should have mentioned earlier that you can test this manually by starting `bitcoind`, waiting for it to start up and sync, and then running:
```
kill -SIGHUP `pgrep bitcoind`
```
Then check `debug.log`; the version information and configuration settings should appear at the end. (You can do that `kill` repeatedly, and the information will appear again each time.)

A more realistic test is is to do actual log rotation: Start `bitcoind`, wait for it to start up, then
```
mv ~/.bitcoin/debug.log ~/.bitcoin/debug.log.old
```
(Note that in this state, additional logging will go to `debug.log.old`; the new `debug.log` won't exist.)

Then send the signal. A new `debug.log` will be created containing the version and configuration information. Further logging will appear there (and not to `debug.log.old`). The intention is that `debug.log.old` can be compressed, or archived (or deleted) or whatever. (See `man logrotate`.)

**Comment by rebroad** [comment]
NACK - does not seem useful or necessary (given the previous logs can be referred to easily).

**Comment by LarryRuane** [comment]
> NACK - does not seem useful or necessary (given the previous logs can be referred to easily).

Well, if it's not useful or necessary, I don't think it's for this reason, because an external log rotation agent (see `logrotate` on Unix-like systems, for example) may have deleted the initial part of the `debug.log` file for the currently-running instance of `bitcoind` that contains the configuration information (which does get logged at startup).

**Comment by rebroad** [comment]
> > NACK - does not seem useful or necessary (given the previous logs can be referred to easily).
> 
> Well, if it's not useful or necessary, I don't think it's for this reason, because an external log rotation agent (see `logrotate` on Unix-like systems, for example) may have deleted the initial part of the `debug.log` file for the currently-running instance of `bitcoind` that contains the configuration information (which does get logged at startup).

ah, I see.. yes, in that case this might be useful, but I think it's better if it's optional - or you could just use a logrotate script to save this part of the debug.log file elsewhere before it deletes the file.

**Comment by LarryRuane** [comment]
Force-pushed rebase, and made a few improvements.

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.
